src/detext/args.py

```diff
@@ -0,0 +1,368 @@
+
+    def _set_late_init_attr(self, attr, value):
+        """Sets an attribute as empty list if it's LateInit"""
+        if getattr(self, attr) is LateInit:
+            setattr(self, attr, value)
+
+    def __post_init__(self):
+        pass
+
+
+@dataclass
+class DatasetArg(Arg):
+    """Dataset related arguments"""
+    distribution_strategy: str = ''  # Distributed training strategy. Reference: tf official models: official/common/distribute_utils.py#L102
+    __distribution_strategy = {'choices': ['one_device', 'mirrored', 'parameter_server', 'multi_worker_mirrored', 'tpu']}
+    all_reduce_alg: Optional[str] = None  # All reduce algorithm. Reference: tf official models: official/common/distribute_utils.py#L102
+    __all_reduce_alg = {'choices': ["hierarchical_copy", "nccl", "ring"]}
+    num_gpu: int = -1  # Number of GPU for training
+
+    run_eagerly: bool = False  # Whether to run in eager mode. Use True for debugging and False for speed
+
+    train_file: str = ''  # Train file.
+    dev_file: str = ''  # Dev file.
+    test_file: str = ''  # Test file.
+    out_dir: str = ''  # Store log/model files.
+
+    num_train_steps: int = 1  # Num steps to train.
+    num_eval_steps: int = 0  # Num steps to eval.
+    num_epochs: int = 0  # Num of epochs to train, will overwrite train_steps if set
+    steps_per_stats: int = 100  # training steps to print statistics.
+    num_eval_rounds: int = 0  # number of evaluation round, this param will override steps_per_eval as max(1, num_train_steps / num_eval_rounds)
+    steps_per_eval: int = 1000  # training steps to evaluate datasets.
+
+    resume_training: bool = False  # Whether to resume training from checkpoint in out_dir.
+    keep_checkpoint_max: int = 5  # The maximum number of recent checkpoint files to keep. If 0, all checkpoint files are kept. Defaults to 5
+
+    train_batch_size: int = 32  # Training data batch size.
+    test_batch_size: int = 32  # Test data batch size.
+
+    def __post_init__(self):
+        super().__post_init__()
+
+        # Check matches between distribution strategy and all reduce algorithm. Reference: tf official models: official/common/distribute_utils.py#L102
+        if self.all_reduce_alg is not None:
+            if self.distribution_strategy == 'mirrored':
+                assert self.all_reduce_alg in [None, 'nccl', 'hierarchical_copy']
+            elif self.distribution_strategy == 'multi_worker_mirrored':
+                assert self.all_reduce_alg in [None, 'nccl', 'ring']
+            else:
+                raise NotImplementedError(
+                    f'Unsupported all reduce algorithm {self.all_reduce_alg} for chosen distribution strategy {self.distribution_strategy}')
+
+        assert self.train_file, 'training_file must be specified'
+        assert self.dev_file, 'dev_file must be specified'
+        assert self.test_file, 'test_file must be specified'
+
+        assert self.out_dir, 'out_dir must be specified'
```

Comment:
line 36-39, if you don't set the initial values, they will be deemed as required. It's probably more expressive that way than assertions.

Reply:
This is set because of this error from Python "SyntaxError: non-default argument follows default argument". As the final DeTextArg is a composition of different arg sub component, I couldn't avoid this pass-around. Any better suggestion?

Reply:
This is set because of this error from Python "SyntaxError: non-default argument follows default argument". As the final DeTextArg is a composition of different arg sub component, I couldn't avoid this pass-around. Any better suggestion?

Reply:
This is an issue of dataclass inheritance. dug around and it seems that no good solutions for python 3.7 yet. A solution was provided for 3.10 https://bugs.python.org/issue43532 . 
Another direction could be: assigning them to dataclass.MISSING, field() and LateInit should work without the manual assertions, but not sure how more readable they are.

Reply:
Thanks for the reference. Nice to know how this relates to the dataclass design.

> assigning them to dataclass.MISSING
Do you mean something like this?
```
a: str = dataclass.MISSING?
```

I'm avoiding using "LateInit" for must have input.